### DISCUSSION TITLE
request_size paramater in libspdm_get_response_finish

### DISCUSSION THREAD

**Discussion by PavelPu123** [DMTF/libspdm#1837] [Q&A]
Hello

As I understand "request_size" in above functions is updated after the transport header is removed, thus the equality sign should stay here, as we shouldn't allow a larger message size (and it should be considered an invalid request). Please correct me if I'm missing something. Thanks.

https://github.com/DMTF/libspdm/blob/4d2b685888740d9632337e4f0c40c61f40d23d7b/library/spdm_responder_lib/libspdm_rsp_finish.c#L453-L454

Commit:
https://github.com/DMTF/libspdm/commit/4d2b685888740d9632337e4f0c40c61f40d23d7b#diff-e20c71e6f0c9740582f8b26119596ac39833a47ed51f0f0da9285e46ad7fab54

**Comment by jyao1** [comment]
We allow the transport layer to add padding. (for example, PCI_DOE requires 4 bytes aligned). A larger message maybe a valid one.
The parser should re-calculate the size. All message parser follows the same pattern.

**Reply by PavelPu123** [reply]
Not sure I get how transport is relevant here. When we're decoding the message the returned size ("decoded_message_size") from:

https://github.com/DMTF/libspdm/blob/51c1cf4ecde9a9b2c7996217c86bebc97a212917/library/spdm_responder_lib/libspdm_rsp_receive_send.c#L182-L185

is:

https://github.com/DMTF/libspdm/blob/51c1cf4ecde9a9b2c7996217c86bebc97a212917/library/spdm_transport_mctp_lib/libspdm_mctp_mctp.c#L152

which is only the size of the data itself. Isn't it?
Thanks.

**Reply by steven-bellock** [reply]
> which is only the size of the data itself. Isn't it?

For some transports that may be true, but not all. In particular I've advocated for getting rid of the checks and getting rid of `message_size` from `libspdm_device_receive_message_func`. See https://github.com/DMTF/libspdm/issues/397

**Reply by PavelPu123** [reply]
I see that both for MCTP and for DOE decoded_message_size is calculated the same:

https://github.com/DMTF/libspdm/blob/51c1cf4ecde9a9b2c7996217c86bebc97a212917/library/spdm_transport_pcidoe_lib/libspdm_doe_pcidoe.c#L191

Do you have an example that supports your claim?

Also, in case the incoming request has additional bytes which are appended after a valid message, shouldn't we alert about this and return an "invalid request"? 

Thanks.

**Reply by steven-bellock** [reply]
Like I said in #397, in the simplest case, suppose there is a shared buffer between the Requester and Responder. The Requester fills the buffer with the request and uses a single bit to alert the Responder that there is a message, whose size is not known to the Responder, in the buffer that needs to be processed. Without any knowledge of the message size from the "transport layer" the Responder can parse the message and provide a response.

**Reply by PavelPu123** [reply]
Ok. 
But again, in case the incoming request has additional bytes which are appended after a valid message (assuming the buffer is large enough), shouldn't we alert about this and return an "invalid request"?

**Reply by steven-bellock** [reply]
Do you assume
1. The receive buffer is first zeroized.
2. The message is placed into the receive buffer.
3. The message is parsed to determine its size.

and if the space outside of the message is non-zero, that this would be considered an invalid request? It's possible that the data outside of the parsed message could be used by an attacker in a [ROP](https://en.wikipedia.org/wiki/Return-oriented_programming)-style attack, especially if the endpoint doesn't support [W^X](https://en.wikipedia.org/wiki/W%5EX). In such a case the endpoint could perform that check, preferably as soon as possible.

**Reply by PavelPu123** [reply]
1. Yes
2. Taking for example KEY_EXCHANGE:

https://github.com/DMTF/libspdm/blob/9ad78c33235296c812e503e84ec7e5a89ba2e19d/library/spdm_responder_lib/libspdm_rsp_key_exchange.c#L167-L171

The receive buffer in this case is "const void *request".

3. The size determination is done here:

https://github.com/DMTF/libspdm/blob/9ad78c33235296c812e503e84ec7e5a89ba2e19d/library/spdm_responder_lib/libspdm_rsp_key_exchange.c#L289-L296

So if "In such a case the endpoint could perform that check, preferably as soon as possible." - where do you propose to perform this check (assuming that message fields are only known when message arrived to this function).

Thanks.

**Reply by PavelPu123** [reply]
@steven-bellock Can you please answer my last question regarding size check? It would be much appreciated. Thanks.

**Reply by steven-bellock** [reply]
The message size that is calculated by parsing would be determined after the `SPDMVersion` and `RequestResponseCode` have been validated.

**Reply by PavelPu123** [reply]
Which is where exactly? Here?
https://github.com/DMTF/libspdm/blob/9ad78c33235296c812e503e84ec7e5a89ba2e19d/library/spdm_responder_lib/libspdm_rsp_key_exchange.c#L200

And how do you propose to do the check (assuming we can't use the "request_size" parameter as mentioned above)?
Thanks.
